### PR TITLE
HDDS-7187. EC: Retry failed writes before rewrite to a new block group

### PR DESCRIPTION
**Pull request by aswinshakil**
## What changes were proposed in this pull request?

When WriteChunk or PutBlock fails when writing a stripe, the client should retry the operation before rewriting this stripe to a new block. So that the block files on the disk will be as large as possible. 

This patch addresses a way to configure `ECXceiverClientGrpc` to retry failed requests based on GRPC Status Code `DEADLINE_EXCEEDED` 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7187

## How was this patch tested?

Manual testing and Unit tests.

### CODE REVIEW

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/ECXceiverClientGrpc.java
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/ECXceiverClientGrpc.java:110]: You could avoid duplicating most of channel creation logic by returning the `channelBuilder` instead.  The EC-specific override could simply call `super.createChannel(...)` and apply the retry-related further customisation to the builder.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientGrpc.java:623]: These become unnecessary with the small refactoring.

File: hadoop-hdds/common/src/main/resources/ozone-default.xml
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:3288]: I'd prefer

 * `ozone.client.ec.grpc.retries.enabled`
 * `ozone.client.ec.grpc.retries.max`

to keep these more closely related.  (Please feel free to drop `grpc` from both if you wish.)

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @aswinshakil for working on this.  Looks good overall, I'd like to suggest some small changes.

File: hadoop-hdds/common/src/main/resources/ozone-default.xml
Comment (aswinshakil) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml:3288]: Thanks for the suggestion. Will update the PR. Adding `grpc` makes sense as this is specific for EC grpc.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/ECXceiverClientGrpc.java
Comment (aswinshakil) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/ECXceiverClientGrpc.java:110]: Thanks for pointing it out. Updating the PR with suggestions.

**Review by adoroszlai — APPROVED** [review]
Thanks @aswinshakil for updating the patch.  I think it's much cleaner now. :+1:

**Comment by sodonnel** [comment]
What will be the result of this change if a node is down for example, and the call fails after a connection timeout sort of error, where the failure doesn't happen immediately? Eg the call fails after 60 seconds. Will it retry 5 times, hanging for 60 seconds each time?

If the DN process is down, but the host is up, usually it will get a connection refused immediately. Will that still retry 5 times?

I guess we are wanting to retry for intermittent failures (eg gc pause), but in the case of a host or process being down, this may lengthen the time for a real failure.

I am not sure what is the best to go on this - there probably isn't a perfect solution.

**Comment by aswinshakil** [comment]
@sodonnel thanks for looking into this. GRPC returns one of the 17 error codes. The error code `UNAVAILABLE` falls under a wider umbrella, where it can also indicate the,
1) GRPC Server is down.
2) Some data is transferred before the connection breaks. 

We want to retry on the second scenario, but not on the first. Since it only sends generic `UNAVAILABLE` it can't be distinguished with GRPC Internal retry policy. 

For the question, Connection refused falls under the `UNAVAILABLE` status code, hence we will retry it. If we are okay with not retrying for the second scenario, We can just keep `DEADLINE_EXCEEDED` for the retry policy. As per the [JIRA comment](https://issues.apache.org/jira/browse/HDDS-7187?focusedCommentId=17598216&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17598216), they have faced only `DEADLINE_EXCEEDED` as per their testing.

**Comment by sodonnel** [comment]
I'm not sure what the best approach is. It might better to start small and handle just the DEADLINE_EXCEEDED and see how it goes in practice. Errors caused by a host down, we probably don't need or want to retry, but as you say it is perhaps impossible to distinguish that error from some others.

**Review by sodonnel — APPROVED** [review]
This LGTM now. We can easily add more codes if needed later now.

@adoroszlai have you any further comments on this last change?

Need to wait for the CI to go green before committing.

File: hadoop-hdds/common/src/main/resources/ozone-default.xml
Comment (umamaheswararao) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:3287]: What is the time out it declares DEADLINE EXCEEDED?
total wait before failing would be: 5 * that_timeout right ?
Should we just set 3? I don't know the reason, but if application just blocked for that much time to get response, then app may behave something wrong? I don't know. 
Can you let me know the total possible wait time?

Comment (aswinshakil) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml:3287]: The client timeout is specified by `ozone.client.read.timeout` which is by default `30s`, After which GRPC throws `DEADLINE_EXCEEDED`. Under heavy load, If we assume we get the same error for the 5 retries, then the wait time would be `5 * 30s = 150s`. 

One solution is that we can make the retry count to 3 (which also includes transient retries grpc tries on its own) so we don't try as much because we don't want to burden the DN that much as well. Or we can also try [throttling the retries](https://github.com/grpc/proposal/blob/master/A6-client-retries.md#throttling-retry-attempts-and-hedged-rpcs)

Comment (umamaheswararao) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml:3287]: Thanks for the details @aswinshakil. Please file a JIRA for metrics or some way we should let people know about these retries happening. I am ok with 3. BTW, I don't think we will burden the DN because, calls may not really reach DN right?

Comment (aswinshakil) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml:3287]: Sure will file a JIRA and update the retry count. Initially, I kept the retry count to 3, thought it might be too small. But makes sense to change it back to 3. 

If the call doesn't even reach the DN, then it won't throw `DEADLINE_EXCEEDED` and rather throw `UNAVAILABLE`, so we should be fine in that case.

Comment (aswinshakil) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml:3287]: Jira for GRPC metrics [HDDS-7530](https://issues.apache.org/jira/browse/HDDS-7530)

**Review by umamaheswararao — APPROVED** [review]
LGTM, pending CI.

**Pull request merged**
